Improve the same user check and its documentation (#69844)
elastic/elasticsearch: x-pack/docs/en/security/limitations.asciidoc

member of `group_2`, only `group_1` will be discovered. However, the
<<active-directory-realm, Active Directory Realm>> *does* support transitive
group membership.

[discrete]
[[same-user-check]]
=== Same user check

The result of <<async-search,async search>> and <<scroll-api,scroll>> can be retrieved later
by the same user who submits the initial request. The same user check is done by comparing
the username, authentication realm type and, for realms other than file or native, realm name.

> Two API keys are never considered as the same user regardless of their privileges.

Author: The same API key, i.e. same API key ID, is treated as the same user. I kinda assumed this implicity when writing the doc. The text was trying to state the opposite side of things, i.e. "Two API keys" are never the same user. I can clarify it a bit more if it does not read right to you.

> Two API keys are never considered as the same user regardless of their privileges.

Reviewer: Sorry for not being clear.
The text does not mention how the same user check works for an API Key and the owner user of the it (it says only about two API keys, not one API key and an user). I know this is hard to explain using the "user" word, so I was suggesting to highlight that an API Key never passes the "same user check" with anyone, except with itself.

> Two API keys are never considered as the same user regardless of their privileges.

Author: Thanks for the clarification. I think Lisa's change has made this point clearer.

This logic also has a few limitations:

* Two different realms can have the same name on different nodes. We do not
recommend configuring realms in this inconsistent way. As such the same user check

> does not attempt to detect this inconsistency.

Author: I added a bullet point for this case.

* The username is computed on the fly for realms backed by certain external authentication
providers. It is in theory possible that two distinct users from the external system get
mapped to the same username. Our recommendation is to avoid this situation in the first place.
Hence the same user check does not account for this potential discrepancy.

> 

Author: That is what I was trying to say in the text. Is it not clear to you? I add a sentence using the DN as an example. Hope it helps.

> 

Reviewer: 🤦 I must've missed it.

member of `group_2`, only `group_1` will be discovered. However, the
<<active-directory-realm, Active Directory Realm>> *does* support transitive
group membership.

[discrete]
[[can-access-resources-check]]
=== Resource sharing check for users and API keys

The result of <<async-search,async search>> and <<scroll-api,scroll>> requests can be retrieved later
by the same user or API key that submitted the initial request. The verification process involves comparing

> the username, authentication realm type, and (for realms other than file, native or kerberos) realm name.

Reviewer 2: ```suggestion
the username, authentication realm type, and (for realms other than file or native) realm name.
```

If you used an API key to submit the request, only that key can retrieve the results.
This logic also has a few limitations:

* Two different realms can have the same name on different nodes. This is not a

> recommended way to configure realms, therefore resource sharing check

Reviewer 3: ```suggestion
recommended way to configure realms, therefore the resource sharing check
```

does not attempt to detect this inconsistency.

> * Realms can be renamed. This can cause inconsistency for resource sharing check

Reviewer 3: ```suggestion
* Realms can be renamed. This can cause inconsistency for the resource sharing check
```

when you submit an async search or scroll then rename the realm and try to retrieve the results.
Hence, changing realm names should be handled with care since it can cause complications for more than
just the resource sharing check.
* The username is dynamically computed for realms backed by certain external authentication

> providers. For example, the username can be derived from part of the DN in a LDAP realm.

Reviewer 3: ```suggestion
providers. For example, the username can be derived from part of the DN in an LDAP realm.
```

It is in theory possible that two distinct users from the external system get
mapped to the same username. Our recommendation is to avoid this situation in the first place.

> Hence resource sharing check does not account for this potential discrepancy.

Reviewer 3: ```suggestion
Hence, the resource sharing check does not account for this potential discrepancy.
```

member of `group_2`, only `group_1` will be discovered. However, the
<<active-directory-realm, Active Directory Realm>> *does* support transitive
group membership.

[discrete]
[[same-user-check]]

> === Same user check

Reviewer: ```Suggestion
=== User and API key same ownership check
```

The result of <<async-search,async search>> and <<scroll-api,scroll>> can be retrieved later
by the same user who submits the initial request. The same user check is done by comparing
the username, authentication realm type and, for realms other than file or native, realm name.
Two API keys are never considered as the same user regardless of their privileges.
This logic also has a few limitations:

* Two different realms can have the same name on different nodes. We do not
recommend configuring realms in this inconsistent way. As such the same user check
does not attempt to detect this inconsistency.
* The same realm can also be renamed. This can cause inconsistency for same user check
when submitting async search/scroll before the rename and retrieving the result after the rename.
. Hence changing realm names should be handled with care
since it can cause complications for more than just same user check.

> * The username is computed on the fly for realms backed by certain external authentication

Reviewer 3: I'm not sure that "on the fly" is clear in all languages. Can we use "dynamically" instead?:
```suggestion
* The username is dynamically computed for realms backed by certain external authentication
```
